**Epoch** = 1 complete forward and backward pass for all training examples

**batch_size** = Number of training samples in 1 batch, i.e during 1 fwd and bwk pass

**No. of iterations** = No. of passes, each pass uses (batch_size) number of samples

Eg. 100 train samples, batch_size=20, 100/20=5 iterations for 1 epoch

In [1]:
import torch
import torchvision
from torch.utils.data import Dataset, DataLoader
import numpy as np
import math
import pandas as pd

### Custom dataset

In [2]:
path = '/content/drive/MyDrive/Datasets/wine.csv'
df = pd.read_csv(path)
print(df.shape)
print("Unique Wine categories = ", df.Wine.unique())
df.head()

(178, 14)
Unique Wine categories =  [1 2 3]


Wine  Alcohol  Malic.acid   Ash   Acl   Mg  Phenols  Flavanoids  \
0     1    14.23        1.71  2.43  15.6  127     2.80        3.06   
1     1    13.20        1.78  2.14  11.2  100     2.65        2.76   
2     1    13.16        2.36  2.67  18.6  101     2.80        3.24   
3     1    14.37        1.95  2.50  16.8  113     3.85        3.49   
4     1    13.24        2.59  2.87  21.0  118     2.80        2.69   

   Nonflavanoid.phenols  Proanth  Color.int   Hue    OD  Proline  
0                  0.28     2.29       5.64  1.04  3.92     1065  
1                  0.26     1.28       4.38  1.05  3.40     1050  
2                  0.30     2.81       5.68  1.03  3.17     1185  
3                  0.24     2.18       7.80  0.86  3.45     1480  
4                  0.39     1.82       4.32  1.04  2.93      735

In [3]:
class WineDataset(Dataset):
    def __init__(self):
        # 1st row is header. So skip it
        xy = np.loadtxt(path, delimiter=",", dtype=np.float32, skiprows=1)
        print("Shape = ", xy.shape)
        self.x = torch.from_numpy(xy[:, 1:]) # Selecting all columns from 1st column onwards
        self.y = torch.from_numpy(xy[:, [0]])
        self.n_samples = xy.shape[0]

    def __getitem__(self, index):
        return self.x[index], self.y[index]

    def __len__(self):
        return self.n_samples

In [4]:
dataset = WineDataset()

first_data = dataset[0] # Calls __getitem__
features, labels = first_data
print(features, labels)
print(f"No. of datasets = {len(dataset)}")

Shape =  (178, 14)
tensor([1.4230e+01, 1.7100e+00, 2.4300e+00, 1.5600e+01, 1.2700e+02, 2.8000e+00,
        3.0600e+00, 2.8000e-01, 2.2900e+00, 5.6400e+00, 1.0400e+00, 3.9200e+00,
        1.0650e+03]) tensor([1.])
No. of datasets = 178


### DataLoader

**num_workers** denotes the number of subprocesses

Since **batch_size=4**, here output will show 4 rows

**shuffle=True**, outputs 4 random rows

In [5]:
batch_size=4
dataloader = DataLoader(dataset=dataset, batch_size=batch_size, shuffle=True, num_workers=2)

dataiter = iter(dataloader)
data = next(dataiter)
features, labels = data
print(f"Features = {features}")
print(f"labels =  {labels}")

Features = tensor([[1.3580e+01, 1.6600e+00, 2.3600e+00, 1.9100e+01, 1.0600e+02, 2.8600e+00,
         3.1900e+00, 2.2000e-01, 1.9500e+00, 6.9000e+00, 1.0900e+00, 2.8800e+00,
         1.5150e+03],
        [1.2220e+01, 1.2900e+00, 1.9400e+00, 1.9000e+01, 9.2000e+01, 2.3600e+00,
         2.0400e+00, 3.9000e-01, 2.0800e+00, 2.7000e+00, 8.6000e-01, 3.0200e+00,
         3.1200e+02],
        [1.2290e+01, 1.4100e+00, 1.9800e+00, 1.6000e+01, 8.5000e+01, 2.5500e+00,
         2.5000e+00, 2.9000e-01, 1.7700e+00, 2.9000e+00, 1.2300e+00, 2.7400e+00,
         4.2800e+02],
        [1.2770e+01, 2.3900e+00, 2.2800e+00, 1.9500e+01, 8.6000e+01, 1.3900e+00,
         5.1000e-01, 4.8000e-01, 6.4000e-01, 9.9000e+00, 5.7000e-01, 1.6300e+00,
         4.7000e+02]])
labels =  tensor([[1.],
        [2.],
        [2.],
        [3.]])


**Here, 44 iterations will process 4 rows each and last iteration will process 178-(4*44) = 2 rows only**

In [6]:
num_epochs = 2
total_samples = len(dataset)
n_iterations_per_epoch = math.ceil(total_samples/batch_size)
print(f"Number of iterations per epoch =  {n_iterations_per_epoch}")

Number of iterations per epoch =  45


In [7]:
# Dummy Training Loop
for epoch in range(1, num_epochs+1):
    print("New Epoch Starts!!!")
    for i, (inputs, labels) in enumerate(dataloader):
        # Forward pass
        if (i+1)%5==0:
            print(f"Epoch {epoch}/{num_epochs}, step {i+1}/{n_iterations_per_epoch},      inputs {inputs.shape}")
    print("#"*100)


New Epoch Starts!!!
Epoch 1/2, step 5/45,      inputs torch.Size([4, 13])
Epoch 1/2, step 10/45,      inputs torch.Size([4, 13])
Epoch 1/2, step 15/45,      inputs torch.Size([4, 13])
Epoch 1/2, step 20/45,      inputs torch.Size([4, 13])
Epoch 1/2, step 25/45,      inputs torch.Size([4, 13])
Epoch 1/2, step 30/45,      inputs torch.Size([4, 13])
Epoch 1/2, step 35/45,      inputs torch.Size([4, 13])
Epoch 1/2, step 40/45,      inputs torch.Size([4, 13])
Epoch 1/2, step 45/45,      inputs torch.Size([2, 13])
####################################################################################################
New Epoch Starts!!!
Epoch 2/2, step 5/45,      inputs torch.Size([4, 13])
Epoch 2/2, step 10/45,      inputs torch.Size([4, 13])
Epoch 2/2, step 15/45,      inputs torch.Size([4, 13])
Epoch 2/2, step 20/45,      inputs torch.Size([4, 13])
Epoch 2/2, step 25/45,      inputs torch.Size([4, 13])
Epoch 2/2, step 30/45,      inputs torch.Size([4, 13])
Epoch 2/2, step 35/45,      inputs t